Update configuration documentation (#11394)
pydantic/pydantic: docs/concepts/config.md

The behaviour of Pydantic can be controlled via a variety of configuration values, documented

> on the [`ConfigDict`][pydantic.ConfigDict] class. This page will describe how configuration
> can be specified on the supported types by Pydantic.

Reviewer: ```suggestion
on the [`ConfigDict`][pydantic.ConfigDict] class. This page describes how configuration can be specified for Pydantic's supported types.
```

## Configuration on Pydantic models

On Pydantic models, configuration can be specified in two ways:

- Using the [`model_config`][pydantic.BaseModel.model_config] class attribute:
  ```python
  from pydantic import BaseModel, ConfigDict, ValidationError

  class Model(BaseModel):
      model_config = ConfigDict(str_max_length=5)  # (1)!

      v: str

  try:
      m = Model(v='abcdef')
  except ValidationError as e:
      print(e)
      """
      1 validation error for Model
      v
        String should have at most 5 characters [type=string_too_long, input_value='abcdef', input_type=str]
      """
  ```

    1. A plain dictionary (i.e. `{'str_max_length': 5}`) can also be used.

    !!! note
        In Pydantic V1, the `Config` class was used. This is still supported, but **deprecated**.

> - Using class arguments:

Reviewer: Should we recommend one over the other?

Author: The benefit of using the class arguments is that it is recognized by type checkers. I think this is only relevant for `frozen`, so actually I'll change the example and add an annotation about it

    """
```

> ## Configuration on type adapters

Reviewer: ```suggestion
## Configuration on `TypeAdapter`s
```

Author: I don't know what's best when it comes to mixing code syntax in titles, and more generally in sentences (like I tend to use _blabla this Pydantic model_ instead of _blabla this `BaseModel`_; I think for type adapter it's fine as you don't actually subclass them but rather use instances of it)

[Type adapters](./type_adapter.md) (using the [`TypeAdapter`][pydantic.TypeAdapter] class) support configuration,

> using the `config` parameter.

Author: That would be _by providing the `config` argument_, but I'll switch to it as I believe _argument_ is understood more easily than _parameter_.

print(m.model_config)
#> {'extra': 'allow', 'str_to_lower': True}
```

> ## Configuration propagation

Reviewer: Good to have this

Note that when using types that support configuration as field annotations on other types,

> configuration will *not* be propagated:

Reviewer: Let's add another explanatory sentence here - basically, config respects model boundaries, or something intuitive like that?
